src/vs/platform/terminal/common/capabilities/shellEnvDetectionCapability.ts

```diff
@@ -0,0 +1,46 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Disposable } from '../../../../base/common/lifecycle.js';
+import { IShellEnvDetectionCapability, TerminalCapability } from './capabilities.js';
+import { Emitter } from '../../../../base/common/event.js';
+
+export class ShellEnvDetectionCapability extends Disposable implements IShellEnvDetectionCapability {
+	readonly type = TerminalCapability.ShellEnvDetection;
+
+	private readonly _env: Map<string, string> = new Map();
+	get env(): Map<string, string> { return this._env; }
+
+	private readonly _onDidChangeEnv = this._register(new Emitter<Map<string, string>>());
+	readonly onDidChangeEnv = this._onDidChangeEnv.event;
+
+	setEnvironment(env: { [key: string]: string | undefined }, isTrusted: boolean): void {
+		if (!isTrusted) {
+			return;
+		}
+
+		let envChanged = false;
+		for (const [key, value] of Object.entries(env)) {
+			if (value !== this._env.get(key)) {
+				envChanged = true;
+				break;
+			}
+		}
```

Comment:
Does `equals` from object.ts work for objects like this? Could save us adding this part